### PR TITLE
Fix: a memory leak in Zlib.set_header

### PR DESCRIPTION
**Pull request by murmour**
Caught this with a static analysis tool while doing a large-scale study of C stub sources on OPAM (as part of [this initiative](https://github.com/ocaml/ocaml/pull/71)). Haven't tested the fix, except for checking that it compiles without warnings or errors. Please, take a careful look at it.

### CODE REVIEW

File: src/zlib_stubs.c [hunk 1]
```diff
@@ -294,8 +299,7 @@ CAMLprim value zlib_set_header(value vstrm, value vheader)
     /* this string is _not_ expected to be zero-terminated */
     len = caml_string_length(Field(Field(vheader,4),0));
     header->extra_len = len;
-    header->extra = caml_stat_alloc(len);
-    memcpy(header->extra, String_val(Field(Field(vheader,4),0)), len);
+    memcpy(wrap->header->extra, String_val(Field(Field(vheader,4),0)), len);
```
Comment (madroach) [inline src/zlib_stubs.c:337]: This whole thing should raise if the buffers are too small:

```
if (caml_string_length() + 1 >= header->extra_len)
   caml_invalid_argument("too long extra string");
```

But maybe keep allocating custom chunks here and fix the finalization procedure.
Also note that header->extra == NULL results in different gzip header than header->extra != NULL and header-> extra_len == 0.

**Comment by madroach** [comment]
Thanks for reporting this!
This needs to be fixed in a slightly different way (see my comments in the diff). I'll have some spare time on Monday.

**Comment by murmour** [comment]
I have updated the patch with the following:
1) Validating the lengths of the provided strings _before_ anything is modified in the header.
2) Assigning `NULL` to the fields of the header that were `None` in the provided structure.

While not particularly elegant, it should do the job. And yes, I agree that using a custom block would be the most simple and robust solution.

**Comment by madroach** [comment]
Thanks for your work and providing patches. I now had a more close look and I'd prefer a solution which uses variable-sized buffers for zlib_set_header(). Also found a buffer over-read and missing call to inflateGetHeader(). This is now  committed. Sorry for not merging your work.
Would you mind doing another run of your static analysis?

**Comment by murmour** [comment]
> Would you mind doing another run of your static analysis?

Sure. I will take a look at the latest version of the code later, using my semi-automatic approach.

By the way, I am an industrial user of your excellent library, so making sure it works well is beneficial for my own software.

**Comment by madroach** [comment]
On Wed, 06 Jan 2016 06:10:32 -0800 Max Mouratov
notifications@github.com wrote:

> > Would you mind doing another run of your static analysis?  
> 
> Sure. I will take a look at the latest version of the code later,
> using my semi-automatic approach.
> 
> By the way, I am an industrial user of your excellent library, so
> making sure it works well is beneficial for my own software.

Wow, "industrial user" of my "hobby project" :-)

I'm not a very experienced C programmer. Do you think the approach I
took to do a single allocation for the gz_header struct and the three
buffers is sensible ?

Christopher

## 

http://gmerlin.de
OpenPGP: http://gmerlin.de/christopher.pub
F190 D013 8F01 AA53 E080  3F3C F17F B0A1 D44E 4FEE

**Comment by murmour** [comment]
> Wow, "industrial user" of my "hobby project" :-)

Oops, pardon me -- I use [camlzip](https://opam.ocaml.org/packages/camlzip/camlzip.1.05/), which is a different wrapper around `zlib`. I use it in production for 3 years, and it never bothered me. Perhaps, you should take a look at its implementation :-)

> I'm not a very experienced C programmer. Do you think the approach I took to do a single allocation for the gz_header struct and the three buffers is sensible ?

I think it's quite confusing and a premature optimization. I'd simply allocate each string in a heap and store them in a similar structure:

``` c
struct wrap_header {
    gz_header zheader;
    char *extra;
    char *name;
    char *comment;
};
```

On finalization, I would check for those `char*` fields being `NULL`.

**Comment by murmour** [comment]
> Oops, pardon me -- I use camlzip, which is a different wrapper around zlib. I use it in production for 3 years, and it never bothered me. Perhaps, you should take a look at its implementation :-)

Your library seems to be a much more complete wrapper, though, so I wish it a prosperous future.

**Comment by murmour** [comment]
> I'd simply allocate each string in a heap and store them in a similar structure.

Actually, an additional structure is not needed, as `gz_header` already has those `char*` fields.

**Comment by madroach** [comment]
Well, sadly that's not the case because zlib may set them to NULL.

**Pull request closed without merging**